### PR TITLE
Loosen the validation rule for URI authority

### LINKED ISSUE #5814: Loosen the validation rule for `Endpoint.host()`
The underscore (`_`) character is not permitted in hostnames but is allowed in other contexts, such as DNS SRV records.

Currently, `URI.parseServerAuthority()` is used to validate a hostname of an `Endpoint` which disallows `_`.
https://github.com/line/armeria/blob/c208353b2edd94df20a1523d7b34bb349c781292/core/src/main/java/com/linecorp/armeria/internal/common/SchemeAndAuthority.java#L53

Armeria internally, `Endpoint.host()` is used for both DNS and a hostname of a URI. So I think we may loosen the validation rule to allow `_` and use it for `DNS` queries and results.

Discord thread: https://discord.com/channels/1087271586832318494/1087272728177942629/1262912254333030451

### PR DESCRIPTION
**Pull request by ikhoon**
Motivation:

The authority part in a URI was validated by
`URI.parseServerAuthority()` which only allows alphanumeric characters, `.` and `-`.
https://github.com/openjdk/jdk/blob/dc35f3e8a84c8f622a4cabb8aee0f96de2e2ea30/src/java.base/share/classes/java/net/URI.java#L3513-L3515

As a result, if underscore (`_`) is set in an authority, `URISyntaxException` is raised. We think the rule is too strict because a request can also be sent to an instance when CSLB is used. `_` is a valid character in a DNS record. Users may want to send a request to a host whose name is `beta_api.cloud.instance-123.somewhere.com`.

Related: #5814

Modifications:

- Remove the usage of `URI.parseServerAuthority()` in `SchemeAndAuthority`
- Parse a hostname and a port from a raw authority.

Result:

- Validation is relaxed to permit underscores (_) in URI's authority.
- Closes #5814

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/internal/common/SchemeAndAuthority.java [hunk 1]
```diff
@@ -18,12 +18,18 @@
...
 import java.net.URI;
 import java.net.URISyntaxException;
 
+import com.google.common.net.HostAndPort;
+
 import com.linecorp.armeria.common.annotation.Nullable;
 
 public final class SchemeAndAuthority {
+
+    private static final char[] RESERVED_CHARS = { '/', '?', '#', };
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/internal/common/SchemeAndAuthority.java:31]: Can we use `BitSet` instead?

File: core/src/main/java/com/linecorp/armeria/internal/common/SchemeAndAuthority.java [hunk 2]
```diff
@@ -48,10 +54,31 @@ public static SchemeAndAuthority of(@Nullable String scheme, String authority) {
...
+            }
+        }
+
         final String authorityWithoutUserInfo = removeUserInfo(authority);
         try {
-            final URI uri = new URI(null, authorityWithoutUserInfo, null, null, null).parseServerAuthority();
-            return new SchemeAndAuthority(scheme, uri.getRawAuthority(), uri.getHost(), uri.getPort());
+            final URI uri = new URI(null, authorityWithoutUserInfo, null, null, null);
+            String rawAuthority = uri.getRawAuthority();
+            if (rawAuthority == null) {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/internal/common/SchemeAndAuthority.java:68]: Can we fail if null or empty? Then we can use `charAt(0) == '['` below.

File: core/src/main/java/com/linecorp/armeria/internal/common/SchemeAndAuthority.java [hunk 3]
```diff
@@ -48,10 +54,31 @@ public static SchemeAndAuthority of(@Nullable String scheme, String authority) {
...
+                throw new IllegalArgumentException("Invalid authority: " + authority);
+            }
+            rawAuthority = IDN.toASCII(rawAuthority, IDN.ALLOW_UNASSIGNED);
+
+            final boolean isIpv6 = rawAuthority.startsWith("[");
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/internal/common/SchemeAndAuthority.java:73]:
```suggestion
            final boolean isIpv6 = rawAuthority.charAt(0) == '[';
```

File: core/src/test/java/com/linecorp/armeria/internal/common/SchemeAndAuthorityTest.java [hunk 4]
```diff
@@ -45,8 +47,15 @@ void fromAuthority(String authority, String expectedAuthority, String expectedHo
...
-            "foo:bar", "http://foo:80", "foo/bar", "foo?bar=1", "foo#bar",
-            "[192.168.0.1]", "[::1", "::1]", "[::1]%eth0", "unix:foo.sock"
+            "foo:bar",        // Invalid port
+            "http://foo:80",  // Scheme included
+            "foo/bar",        // Authority with path
+            "foo?bar=1",      // Authority with query
+            "foo#bar",        // Authority with fragment
+            "[192.168.0.1]",  // Bracketed IPv4
+            "[::1", "::1]",   // Incomplete IPv6
+            "[::1]%eth0",     // IPv6 with scope
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/internal/common/SchemeAndAuthorityTest.java:57]: I think an IPv6 address with scope is represented like this:

- `[::1%eth0]:8080`

In this case, we should simply strip the scope and normalize into:

- `[::1]:8080`

However, `[::1]%eth0` should remain failing.

File: core/src/test/java/com/linecorp/armeria/internal/common/SchemeAndAuthorityTest.java [hunk 5]
```diff
@@ -25,13 +25,15 @@
...
     @CsvSource({
-            "0.0.0.0:80,        0.0.0.0:80,         0.0.0.0,            80",    // IPv4
-            "[::1]:8080,        [::1]:8080,         [::1],              8080",  // IPv6
-            "unix%3Afoo.sock,   unix%3Afoo.sock,    unix%3Afoo.sock,    -1",    // Domain socket
-            "foo.bar,           foo.bar,            foo.bar,            -1",    // Only host
-            "foo:,              foo:,               foo,                -1",    // Empty port
-            "bar:80,            bar:80,             bar,                80",    // Host and port
-            "foo@bar:80,        bar:80,             bar,                80",    // Userinfo and host and port
+            "0.0.0.0:80,        0.0.0.0:80,          0.0.0.0,            80",    // IPv4
+            "[::1]:8080,        [::1]:8080,          [::1],              8080",  // IPv6
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/internal/common/SchemeAndAuthorityTest.java:29]: I realize that we don't have the cases for the IP addresses without port:

```suggestion
            "0.0.0.0,           0.0.0.0,             0.0.0.0,            -1",    // IPv4
            "0.0.0.0:,          0.0.0.0:,            0.0.0.0,            -1",    // IPv4 with empty port
            "0.0.0.0:80,        0.0.0.0:80,          0.0.0.0,            80",    // IPv4 with port 
            "[::1],             [::1],               [::1],              -1",    // IPv6
            "[::1]:,            [::1]:,              [::1],              -1",    // IPv6 with empty port
            "[::1]:8080,        [::1]:8080,          [::1],              8080",  // IPv6 with port
```

File: core/src/test/java/com/linecorp/armeria/internal/common/SchemeAndAuthorityTest.java (hunk 4 above)
Comment (ikhoon) [reply core/src/test/java/com/linecorp/armeria/internal/common/SchemeAndAuthorityTest.java:57]: TIL 📝

**Review by minwoox — APPROVED** [review]
👍 👍 👍

File: core/src/test/java/com/linecorp/armeria/internal/common/SchemeAndAuthorityTest.java [hunk 6]
```diff
@@ -45,8 +52,16 @@ void fromAuthority(String authority, String expectedAuthority, String expectedHo
...
     @CsvSource({
-            "foo:bar", "http://foo:80", "foo/bar", "foo?bar=1", "foo#bar",
-            "[192.168.0.1]", "[::1", "::1]", "[::1]%eth0", "unix:foo.sock"
+            "foo:bar",                 // Invalid port
+            "http://foo:80",           // Scheme included
+            "foo/bar",                 // Authority with path
+            "foo?bar=1",               // Authority with query
+            "foo#bar",                 // Authority with fragment
+            "[192.168.0.1]",           // Bracketed IPv4
+            "[::1", "::1]",            // Incomplete IPv6 with scope
```
Comment (ikhoon) [inline core/src/test/java/com/linecorp/armeria/internal/common/SchemeAndAuthorityTest.java:61]:
```suggestion
            "[::1", "::1]",            // Incomplete IPv6
```

File: core/src/test/java/com/linecorp/armeria/internal/common/SchemeAndAuthorityTest.java [hunk 7]
```diff
@@ -25,13 +25,20 @@
...
-            "foo:,              foo:,               foo,                -1",    // Empty port
-            "bar:80,            bar:80,             bar,                80",    // Host and port
-            "foo@bar:80,        bar:80,             bar,                80",    // Userinfo and host and port
+            "0.0.0.0,           0.0.0.0,             0.0.0.0,            -1",    // IPv4
+            "0.0.0.0:,          0.0.0.0:,            0.0.0.0,            -1",    // IPv4 with empty port
+            "0.0.0.0:80,        0.0.0.0:80,          0.0.0.0,            80",    // IPv4 with port
+            "[::1],             [::1],               [::1],              -1",    // IPv6
+            "[::1]:,            [::1]:,              [::1],              -1",    // IPv6 with empty port
+            "[::1]:8080,        [::1]:8080,          [::1],              8080",  // IPv6 with port
+            "[::1%eth0]:8080,   [::1]:8080,          [::1],              8080",  // IPv6 with port
```
Comment (ikhoon) [inline core/src/test/java/com/linecorp/armeria/internal/common/SchemeAndAuthorityTest.java:34]:
```suggestion
            "[::1%eth0]:8080,   [::1]:8080,          [::1],              8080",  // IPv6 with port and scope
```

**Pull request merged**